File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
// link :https://leetcode.com/problems/find-median-from-data-stream/discuss/1506078/Java-Detailed-Code-Solutions-for-Follow-Ups



/**
 * Using two heaps
 * 
 * Time Complexity:
 * 1) addNum -> O(5 * log (N/2)) = O(log N)
 * 2) findMedian -> O(1)
 *
 * Space Complexity: O(N)
 *
 * N = Count of numbers in the data stream.
 */
class MedianFinder {
    PriorityQueue<Integer> smallNums; // Max Heap
    PriorityQueue<Integer> largeNums; // Min Heap

    public MedianFinder() {
        smallNums = new PriorityQueue<>(Collections.reverseOrder());
        largeNums = new PriorityQueue<>();
    }

    public void addNum(int num) {
        // Add to maxHeap first (Group of smaller numbers)
        smallNums.offer(num);
        // Balance the heaps
        largeNums.offer(smallNums.poll());
        if (largeNums.size() > smallNums.size()) {
            smallNums.offer(largeNums.poll());
        }
    }

    public double findMedian() {
        if (smallNums.size() != largeNums.size()) {
            return smallNums.peek();
        }
        return (smallNums.peek() + largeNums.peek()) / 2.0;
    }
}

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.py
import heapq
class MedianFinder:

    def __init__(self):
        self.min_heap=[]
        self.max_heap=[]

    def addNum(self, num: int) -> None:
        heapq.heappush(self.min_heap,-num)

        if self.min_heap and self.max_heap and -self.min_heap[0]>self.max_heap[0]:
            heapq.heappush(self.max_heap,-heapq.heappop(self.min_heap))
        # Balance sizes
        if len(self.min_heap)>len(self.max_heap)+1:
            heapq.heappush(self.max_heap,-heapq.heappop(self.min_heap))
        elif len(self.max_heap)>len(self.min_heap):
            heapq.heappush(self.min_heap,-heapq.heappop(self.max_heap))
    def findMedian(self) -> float:
        if len(self.min_heap)>len(self.max_heap):
            return -self.min_heap[0]
        else:
            return (-self.min_heap[0]+self.max_heap[0])/2
        
# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()